saw-core/src/Verifier/SAW/Simulator/TermModel.hs

```diff
@@ -0,0 +1,1092 @@
+
+type ReplaceUninterpMap = Map VarIndex [(ExtCns Term, [Term])]
+
+
+extractUninterp ::
+  (?recordEC :: BoundECRecorder) =>
+  SharedContext ->
+  ModuleMap ->
+  Map Ident TmPrim {- ^ additional primitives -} ->
+  Map VarIndex TmValue {- ^ ExtCns values -} ->
+  Set VarIndex {- ^ 'unints' Constants in this set are kept uninterpreted -} ->
+  Set VarIndex {- ^ 'opaque' Constants in this set are not evaluated -} ->
+  Term ->
+  IO (Term, ReplaceUninterpMap)
+extractUninterp sc m addlPrims ecVals unintSet opaqueSet t =
+  do mapref <- newIORef mempty
+     cfg <- mfix (\cfg -> Sim.evalGlobal' m (Map.union (constMap sc cfg) addlPrims)
+                             (extcns cfg mapref) (uninterpreted cfg mapref) (neutral cfg) (primHandler cfg))
+     v <- Sim.evalSharedTerm cfg t
+     tv <- evalType cfg =<< scTypeOf sc t
+     t' <- readBackValue sc cfg tv v
+     replMap <- readIORef mapref
+     return (t', replMap)
+
+ where
+    uninterpreted cfg mapref tf ec@(EC ix _nm ty)
+      | Set.member ix opaqueSet = Just $
+          do tm <- scTermF sc tf
+             reflectTerm sc cfg ty tm
+      | Set.member ix unintSet = Just (replace sc cfg mapref ec)
+      | otherwise = Nothing
+
+    extcns cfg mapref tf ec@(EC ix _nm ty)
+      | Set.member ix unintSet = replace sc cfg mapref ec
+      | otherwise =
+          case Map.lookup ix ecVals of
+            Just v  -> return v
+            Nothing ->
+              do tm <- scTermF sc tf
+                 reflectTerm sc cfg ty tm
+
+    neutral cfg env nt =
+      do env' <- traverse (\(x,ty) -> readBackValue sc cfg ty =<< force x) env
+         tm   <- instantiateVarList sc 0 env' =<< neutralToSharedTerm sc nt
+         tyv  <- evalType cfg =<< scTypeOf sc tm
+         reflectTerm sc cfg tyv tm
+
+    primHandler cfg pn _msg env tp =
+      do pn'  <- traverse (readBackTValue sc cfg) pn
+         args <- reverse <$> traverse (\(x,ty) -> readBackValue sc cfg ty =<< force x) env
+         prim <- scFlatTermF sc (Primitive pn')
+         f    <- foldM (scApply sc) prim args
+         reflectTerm sc cfg tp f
+
+replace ::
+  (?recordEC :: BoundECRecorder) =>
+  SharedContext ->
+  Sim.SimulatorConfig TermModel ->
+  IORef ReplaceUninterpMap ->
+  ExtCns (TValue TermModel) ->
+  IO (Value TermModel)
+replace sc cfg mapref ec = loop [] (ecType ec)
+  where
+    loop :: [Term] -> TValue TermModel -> IO (Value TermModel)
+    loop env (VPiType nm xty body) =
+        return $ VFun nm $ \x ->
+          do ty <- applyPiBody body x
+             xtm <- readBackValue sc cfg xty =<< force x
+             loop (xtm:env) ty
+
+    loop env ty =
+      do let args = reverse env
+         ty' <- readBackTValue sc cfg ty
+         newec <- scFreshEC sc (toShortName (ecName ec)) ty'
+         modifyIORef mapref (Map.alter (Just . ((newec,args):) . fromMaybe []) (ecVarIndex ec))
+         reflectTerm sc cfg ty =<< scFlatTermF sc (ExtCns newec)
+
+normalizeSharedTerm ::
```

Comment:
I wonder whether there are any useful abstractions that could be shared between this and `extractUninterp`. It's probably not critical to do so, and might not even be helpful, but it could be worth considering.

Reply:
There might be, but I think it's actually clearer with it broken out separately like this.